File: 414.ThirdMaximumNumber/max.cc
#include <climits>

class Solution {
public:
    int thirdMax(vector<int>& nums) {
        long max1 = LONG_MIN, max2 = LONG_MIN, max3 = LONG_MIN;
        for (int i = 0; i < int(nums.size()); i++) {
            if (nums[i] > max1) {
                max3 = max2;
                max2 = max1;
                max1 = nums[i];
            }
            else if (nums[i] == max1) {
                continue;
            }
            else if (nums[i] > max2) {
                max3 = max2;
                max2 = nums[i];
            }
            else if (nums[i] == max2) {
                continue;
            }
            else if (nums[i] > max3) {
                max3 = nums[i];
            }
        }

        if (max3 > LONG_MIN) {
            return int(max3);
        }
        return int(max1);
    }
};

File: 414.ThirdMaximumNumber/max.py
class Solution(object):
    def thirdMax(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        max1 = max2 = max3 = -float('inf')
        for n in nums:
            if n > max1:
                max3, max2, max1 = max2, max1, n
            elif n == max1:
                continue
            elif n > max2:
                max3, max2 = max2, n
            elif n == max2:
                continue
            elif n > max3:
                max3 = n

        if max3 > -float('inf'):
            return max3
        return max1


if __name__ == '__main__':
    nums = [3,3,4,3,4,3,0,3,3]
    print Solution().thirdMax(nums)
